File: web/src/algorithms/snippets/bitonic.java
import java.util.Arrays;

/** Bitonic sort — pads to the next power of two. */
public final class BitonicSort {
    public static void sort(int[] data) {
        int n = data.length;
        if (n < 2) return;
        int m = 1;
        while (m < n) m <<= 1;
        if (m == n) {
            bitonicPow2(data, n);
            return;
        }
        int[] buf = new int[m];
        Arrays.fill(buf, Integer.MAX_VALUE);
        System.arraycopy(data, 0, buf, 0, n);
        bitonicPow2(buf, m);
        System.arraycopy(buf, 0, data, 0, n);
    }

    private static void bitonicPow2(int[] a, int n) {
        for (int k = 2; k <= n; k <<= 1) {
            for (int j = k >> 1; j > 0; j >>= 1) {
                for (int i = 0; i < n; i++) {
                    int l = i ^ j;
                    if (l <= i) continue;
                    boolean asc = (i & k) == 0;
                    if ((asc && a[i] > a[l]) || (!asc && a[i] < a[l])) {
                        int t = a[i];
                        a[i] = a[l];
                        a[l] = t;
                    }
                }
            }
        }
    }
}

File: web/src/algorithms/snippets/bitonic.py
"""Bitonic sort — pad to next power of two with a large sentinel."""

def bitonic_sort(data: list[int]) -> list[int]:
    n = len(data)
    if n < 2:
        return data[:]
    m = 1
    while m < n:
        m <<= 1
    a = data[:] + [2**31 - 1] * (m - n)
    _bitonic_pow2(a, m)
    return a[:n]


def _bitonic_pow2(a: list[int], n: int) -> None:
    k = 2
    while k <= n:
        j = k >> 1
        while j > 0:
            for i in range(n):
                l = i ^ j
                if l <= i:
                    continue
                asc = (i & k) == 0
                if (asc and a[i] > a[l]) or (not asc and a[i] < a[l]):
                    a[i], a[l] = a[l], a[i]
            j >>= 1
        k <<= 1
